.github/workflows/environment.yml

```diff
@@ -17,11 +17,11 @@ jobs:
     env:
       ENVIRONMENT_AUTOBUILD: true
     steps:
-      - run: docker login https://docker.pkg.github.com -u ${{ github.actor }} -p ${{ secrets.GITHUB_TOKEN }}
+      - run: docker login -u ${{ secrets.DOCKER_USER }} -p ${{ secrets.DOCKER_PASSWORD }}
       - uses: actions/checkout@v2
       - run: make ci-sweep
       - name: prepare environment
         run: make environment-prepare
       - name: push image
         if: github.event_name != 'workflow_dispatch'
-        run: docker push docker.pkg.github.com/timberio/vector/environment:${{ github.sha }}
+        run: docker push timberio/ci_image:${{ github.sha }}
```

Comment:
What's the idea behind using the `${{ github.sha }}`? I couldn't figure it out before, and I don't get it now :(

If jobs that build and consume the image start simultaneously, and the image if scoped to the commit sha - how can it be possibly available in time for the consuming job to not fail?

I think we should either order the environment jobs before the rest of the jobs via `needs` (and move them to the same workflow file), or use some other way to stall the kick-off of the jobs that consume the image until the image is available.

A decent alternative would be to build the environment images nightly and push with a sliding tag `nightly`. Then at the consumer location, allow using the image tagged with the branch name, or, if no branch-specific image was pushed - fallback to `nightly`. This was people will be able to push their own custom `ci_image` tags with per-branchs specifics. We can also allow them building them in the cloud, via manually triggering the environment workflow via `workflow_dispatch` event, allowing to build an image off of a specific branch, and push it with the branch name instead of `nightly`.

Reply:
That's exactly what we want to avoid by using a container - the time needed to bootstrap or in the case you're proposing to build the container. Github Actions will pull `timberio/ci_image:latest`. This is the last image built off `master` by the `environment` workflow. I think it's very unlikely a CI job will fail because the `ci_image` is out of date. Everything vector-related gets added by `cargo` or elsewhere.

Reply:
Good, I see how what you explained above can work.

But I'm still not getting why, instead of pushing `timberio/ci_image:latest` here, we push `timberio/ci_image:${{ github.sha }}`.

Reply:
Largely because if we have a bad image we can easily roll back by deleting that tag without needing to re-roll the image.

Reply:
But docker doesn’t make these images accessible as `:latest`, does it? There’s no automatic rule there on a server side to exposed the last uploaded image with the `:latest` tag, is there?